File: lib/models/product.dart
class Product {
  final int productId;
  final int supplierId;
  final String productName;
  final String productCode;
  final String sku;
  final double unitCost;
  final int currentQty;
  final int? leadTimeDays;
  final int safetyStock;
  final int reorderPoint;
  final String? serialNo;
  final String statusFlag;

  Product({
    required this.productId,
    required this.supplierId,
    required this.productName,
    required this.productCode,
    required this.sku,
    required this.unitCost,
    required this.currentQty,
    this.leadTimeDays,
    required this.safetyStock,
    required this.reorderPoint,
    this.serialNo,
    required this.statusFlag,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    Map<String, dynamic>? supplierData;
    final supplier = json['supplier'];
    if (supplier is List && supplier.isNotEmpty && supplier.first is Map) {
      supplierData = supplier.first as Map<String, dynamic>;
    } else if (supplier is Map<String, dynamic>) {
      supplierData = supplier;
    }

    Map<String, dynamic>? reorderParameter;
    final relation = json['reorder_parameter'];
    if (relation is List && relation.isNotEmpty && relation.first is Map) {
      reorderParameter = relation.first as Map<String, dynamic>;
    } else if (relation is Map<String, dynamic>) {
      reorderParameter = relation;
    }

    int? intValue(dynamic value) {
      if (value is num) {
        return value.toInt();
      }
      return int.tryParse(value?.toString() ?? '');
    }

    String statusValue = json['status_flag']?.toString() ?? 'In Stock';
    switch (statusValue.toUpperCase()) {
      case 'LOW_STOCK':
        statusValue = 'Low Stock';
        break;
      case 'OVERSTOCK':
        statusValue = 'High Stock';
        break;
      case 'DISCONTINUED':
        statusValue = 'Discontinued';
        break;
    }

    return Product(
      productId: intValue(json['product_id'] ?? json['id']) ?? 0,
      supplierId:
          intValue(json['supplier_id'] ?? supplierData?['supplier_id'] ?? supplierData?['id']) ??
          0,
      productName: (json['product_name'] ?? json['name'] ?? '').toString(),
      productCode: (json['product_code'] ?? '').toString(),
      sku: (json['sku'] ?? '').toString(),
      unitCost: double.tryParse((json['unit_cost'] ?? 0).toString()) ?? 0,
      currentQty: intValue(json['current_qty']) ?? 0,
      leadTimeDays:
          intValue(json['lead_time_days']) ??
          intValue(reorderParameter?['lead_time_days']) ??
          intValue(supplierData?['lead_time_days']),
      safetyStock:
          intValue(json['safety_stock']) ??
          intValue(json['overstock_level']) ??
          intValue(reorderParameter?['safety_stock']) ??
          0,
      reorderPoint: intValue(json['reorder_point'] ?? json['reorder_level']) ?? 0,
      serialNo: json['serial_no']?.toString(),
      statusFlag: statusValue,
    );
  }
}

File: app/models/product.py
from decimal import Decimal
from sqlalchemy import Enum, ForeignKey, Integer, Numeric, String
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.models.base import Base


class Product(Base):
    __tablename__ = "product"

    id: Mapped[int] = mapped_column("product_id", Integer, primary_key=True, index=True)
    product_code: Mapped[str] = mapped_column(String(50), unique=True, nullable=False, index=True)
    sku: Mapped[str] = mapped_column(String(50), unique=True, nullable=False, index=True)
    name: Mapped[str] = mapped_column("product_name", String(100), nullable=False, index=True)
    supplier_id: Mapped[int] = mapped_column(Integer, ForeignKey("supplier.supplier_id"), nullable=False, index=True)
    current_qty: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    reorder_level: Mapped[int] = mapped_column("reorder_point", Integer, default=0, nullable=False)
    unit_cost: Mapped[Decimal] = mapped_column(Numeric(12, 2), default=Decimal("0.00"), nullable=False)
    status_flag: Mapped[str | None] = mapped_column(
        Enum("Low Stock", "In Stock", "High Stock", "Discontinued", name="product_status"), 
        nullable=True, 
        index=True
    )
    serial_no: Mapped[int | None] = mapped_column(Integer, nullable=True)

    supplier: Mapped["Supplier"] = relationship("Supplier", back_populates="products")
    stock_receipts: Mapped["list[StockReceipt]"] = relationship("StockReceipt", back_populates="product")
    reorder_params: Mapped["ReorderParameter | None"] = relationship("ReorderParameter", back_populates="product", uselist=False)

    @property
    def overstock_level(self) -> int:
        return self.reorder_level

    @property
    def lead_time_days(self) -> int:
        return self.reorder_params.lead_time_days if self.reorder_params else 0
